stdlib/public/stubs/MathStubs.cpp

```diff
@@ -0,0 +1,146 @@
+//===--- MathStubs.cpp - Swift Language Runtime Stubs ---------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Math stubs for functions which should be defined in the core standard
+// library, but are difficult or impossible to write in Swift at the
+// moment.
+//
+//===----------------------------------------------------------------------===//
+
+#include "../SwiftShims/Visibility.h"
+
+#include <climits>
+#include <cstdlib>
+
+#if __has_attribute(__mode__)
+#define SWIFT_MODE_DI __attribute__((__mode__(DI)))
+#define SWIFT_MODE_TI __attribute__((__mode__(TI)))
+#else
+#define SWIFT_MODE_DI
+#define SWIFT_MODE_TI
+#endif
+
+typedef int di_int SWIFT_MODE_DI;
+typedef int ti_int SWIFT_MODE_TI;
+
+extern "C" {
+
+// Although this builtin is provided by clang rt builtins,
+// it isn't provided by libgcc, which is the default
+// runtime library on Linux, even when compiling with clang.
+// This implementation is copied here to avoid a new dependency
+// on compiler-rt on Linux.
+// FIXME: rdar://14883575 Libcompiler_rt omits muloti4
```

Comment:
This is really unfortunate.  Is avoiding the compiler_rt dependency important for some reason?

Reply:
That was pre-existing.  It is relatively nice to avoid the compiler-rt dependency as it causes a lot of headaches on Linux. libgcc is the more common compiler runtime, and it also includes an unwinder.  On Windows, it also collides with another (static) library - `vcclang.lib`.

Reply:
Note that I didn't rebase the branch, so that commit is not really part of the change here.